### PR TITLE
macOS build

### LINKED ISSUE #1: Add support for building under macOS
I tried to compile the fileformat project under macOS, but failed.

first. the llvm project link rt and tinfo lib, need remove it, and set `LLVM_ENABLE_TERMINFO` to false. 

second. the openssl project need to add macOS support. i added `OS_MACOSX` var to setting the `OPENSSL_CONFIGURE_ARCH` to `darwin64-x86_64-cc`.

third. the src/ar-extractor/archive_wrapper.cpp file, need a little modification.

```std::size_t counter = 0;``` to ```uint64_t counter = 0;```

after all of the above operations are completed, the compilation process still fails, returning the following error:

```
[ 41%] Linking CXX executable ar-extractor
[ 41%] Linking CXX executable macho-extractor
[ 41%] Building CXX object src/fileformat/CMakeFiles/fileformat.dir/file_format/intel_hex/intel_hex_format.cpp.o
[ 41%] Built target macho-extractortool
[ 41%] Built target ar-extractortool
In file included from /Users/android/git/fileformat/src/fileformat/file_format/intel_hex/intel_hex_format.cpp:7:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:702:71: error: invalid operands to binary expression
      ('const fileformat::IntelHexSection' and 'const fileformat::IntelHexSection')
    bool operator()(const _T1& __x, const _T1& __y) const {return __x < __y;}
                                                                  ~~~ ^ ~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:3855:17: note: in instantiation of member function
      'std::__1::__less<fileformat::IntelHexSection, fileformat::IntelHexSection>::operator()' requested here
            if (__comp(*--__last, *__first))
                ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:4041:5: note: in instantiation of function template
      specialization 'std::__1::__sort<std::__1::__less<fileformat::IntelHexSection, fileformat::IntelHexSection> &, fileformat::IntelHexSection *>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:4050:12: note: in instantiation of function template
      specialization 'std::__1::sort<fileformat::IntelHexSection *, std::__1::__less<fileformat::IntelHexSection, fileformat::IntelHexSection> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:4066:12: note: in instantiation of function template
      specialization 'std::__1::sort<fileformat::IntelHexSection *>' requested here
    _VSTD::sort(__first.base(), __last.base());
           ^
/Users/android/git/fileformat/src/fileformat/file_format/intel_hex/intel_hex_format.cpp:77:7: note: in instantiation of function template specialization
      'std::__1::sort<fileformat::IntelHexSection>' requested here
        std::sort(parser.sections.begin(), parser.sections.end());
             ^
/Users/android/git/fileformat/include/fileformat/file_format/intel_hex/intel_hex_parser/intel_hex_parser.h:34:8: note: candidate function not viable: 'this' argument has
      type 'const fileformat::IntelHexSection', but method is not marked const
                bool operator<(IntelHexSection const &a);
                     ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:430:1: note: candidate template ignored: could not match
      'pair<type-parameter-0-0, type-parameter-0-1>' against 'const fileformat::IntelHexSection'
operator< (const pair<_T1,_T2>& __x, const pair<_T1,_T2>& __y)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/iterator:592:1: note: candidate template ignored: could not match
      'reverse_iterator<type-parameter-0-0>' against 'const fileformat::IntelHexSection'
operator<(const reverse_iterator<_Iter1>& __x, const reverse_iterator<_Iter2>& __y)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/iterator:996:1: note: candidate template ignored: could not match
      'move_iterator<type-parameter-0-0>' against 'const fileformat::IntelHexSection'
operator<(const move_iterator<_Iter1>& __x, const move_iterator<_Iter2>& __y)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/iterator:1312:1: note: candidate template ignored: could not
      match '__wrap_iter<type-parameter-0-0>' against 'const fileformat::IntelHexSection'
operator<(const __wrap_iter<_Iter1>& __x, const __wrap_iter<_Iter2>& __y) _NOEXCEPT
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/tuple:952:1: note: candidate template ignored: could not match
      'tuple<type-parameter-0-0...>' against 'const fileformat::IntelHexSection'
operator<(const tuple<_Tp...>& __x, const tuple<_Up...>& __y)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2995:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const fileformat::IntelHexSection'
operator< (const unique_ptr<_T1, _D1>& __x, const unique_ptr<_T2, _D2>& __y)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3053:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const fileformat::IntelHexSection'
operator<(const unique_ptr<_T1, _D1>& __x, nullptr_t)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3062:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const fileformat::IntelHexSection'
operator<(nullptr_t, const unique_ptr<_T1, _D1>& __x)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:4889:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const fileformat::IntelHexSection'
operator<(const shared_ptr<_Tp>& __x, const shared_ptr<_Up>& __y) _NOEXCEPT
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:4954:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const fileformat::IntelHexSection'
operator<(const shared_ptr<_Tp>& __x, nullptr_t) _NOEXCEPT
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:4962:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const fileformat::IntelHexSection'
operator<(nullptr_t, const shared_ptr<_Tp>& __x) _NOEXCEPT
^
In file included from /Users/android/git/fileformat/src/fileformat/file_format/intel_hex/intel_hex_format.cpp:7:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:3642:20: error: no matching function for call to
      '__sort3'
    unsigned __r = __sort3<_Compare>(__x1, __x2, __x3, __c);
```

here is the patch:
```
$ git format-patch -1
$ cat 0001-add-macOS-support.patch
From 5a87a0d5d55a162b93572c33e04b5024037b6e6b Mon Sep 17 00:00:00 2001
From: rmbp <fei_cong.hotmail.com>
Date: Wed, 13 Dec 2017 13:59:14 +0800
Subject: [PATCH] add macOS support.

---
 deps/openssl/CMakeLists.txt          | 19 ++++++++++++++++++-
 src/ar-extractor/archive_wrapper.cpp |  2 +-
 2 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/deps/openssl/CMakeLists.txt b/deps/openssl/CMakeLists.txt
index afa11d3..dd77e2f 100644
--- a/deps/openssl/CMakeLists.txt
+++ b/deps/openssl/CMakeLists.txt
@@ -15,6 +15,8 @@ endif()
 
 if(WIN32)
 	set(OS_WINDOWS 1)
+elseif(APPLE)
+	set(OS_MACOSX 1)
 elseif(UNIX)
 	set(OS_LINUX 1)
 else()
@@ -27,7 +29,15 @@ if(OS_WINDOWS)
 	else()
 		message(ERROR "Unsupported compiler. Supported: MSVC.")
 	endif()
-else(OS_LINUX)
+elseif(OS_MACOSX)
+	if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+		set(COMPILER_GCC 1)
+	elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+		set(COMPILER_CLANG 1)
+	else()
+		set(COMPILER_CLANG 1)
+	endif()
+elseif(OS_LINUX)
 	if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
 		set(COMPILER_GCC 1)
 	elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
@@ -46,6 +56,13 @@ if(OS_WINDOWS)
 			set(OPENSSL_CONFIGURE_ARCH "VC-WIN64A")
 		endif()
 	endif()
+elseif(OS_MACOSX)
+	set(OPENSSL_MAKE make -j${CPUS})
+	if(ARCH_32BIT)
+		set(OPENSSL_CONFIGURE_ARCH "darwin-i386-cc")
+	elseif(ARCH_64BIT)
+		set(OPENSSL_CONFIGURE_ARCH "darwin64-x86_64-cc")
+	endif()
 elseif(OS_LINUX)
     set(OPENSSL_MAKE make -j${CPUS})
 	if(ARCH_32BIT)
diff --git a/src/ar-extractor/archive_wrapper.cpp b/src/ar-extractor/archive_wrapper.cpp
index 3a654eb..251425c 100644
--- a/src/ar-extractor/archive_wrapper.cpp
+++ b/src/ar-extractor/archive_wrapper.cpp
@@ -213,7 +213,7 @@ bool ArchiveWrapper::getJsonList(
 	bool numbers) const
 {
 	result.clear();
-	std::size_t counter = 0;
+	uint64_t counter = 0;
 	std::vector<std::string> names;
 
 	if (getNames(names, errorMessage)) {
-- 
2.15.1

```

how can I solve this problem?

### PR DESCRIPTION
**Pull request by stek29**
Fixes #1 
Needs updated llvm submodule (see avast-tl/llvm#1)

### CODE REVIEW

**Comment by stek29** [comment]
builds successfully now :D

**Comment by metthal** [comment]
I don't have Mac machine, but if this works for you I'll merge it after I make sure that it builds also on Linux and Windows. I'll hopefully get to it today.

File: deps/openssl/CMakeLists.txt [hunk 1]
```diff
@@ -15,10 +15,12 @@ endif()
 
 if(WIN32)
 	set(OS_WINDOWS 1)
-elseif(UNIX)
+elseif(LINUX)
```
Comment (metthal) [inline deps/openssl/CMakeLists.txt]: This line breaks Linux build. Variable `LINUX` is not defined by CMake. CMake defines only `UNIX` but Mac OS X also falls under `UNIX` category.

What I propose is that we will rather use [CMAKE_SYSTEM_NAME](https://cmake.org/cmake/help/v3.6/variable/CMAKE_SYSTEM_NAME.html) in order to detect OS. Unfortunately, documentation does not list all possible values this variable can contain. I've tested it on Linux and Windows and it  contained `Linux` and `Windows`, respectively.

Could you please test out what value is there on Mac and update the patch? Thank you.

**Comment by stek29** [comment]
@metthal Can you try again?

**Comment by metthal** [comment]
I merged your changes and added some other changes from https://github.com/avast-tl/retdec/issues/7 (patch by @mabrowning). So fileformat should be compilable under Mac OS X.

**Pull request merged**